darts/models/forecasting/regression_model.py

```diff
@@ -582,123 +463,114 @@ def predict(
-                    "are requires to be at least `n + max_lags`, with `max_lag` being the furthest lag in the future"
-                    f"For the {sample}-th sample, last future covariate timestamp is"
-                    f"{future_covariates[sample].end_time()}, whereas it should be at least {last_req_ts}.",
-                )
+        # prediction preprocessing
+
+        covariates = {
+            "past": (past_covariates, self.lags.get("past")),
+            "future": (future_covariates, self.lags.get("future")),
+        }
+
+        # dictionary containing covariate data over time span required for prediction
+        covariate_matrices = {}
+        # dictionary containing covariate lags relative to minimum covariate lag
+        relative_cov_lags = {}
+        # number of prediction steps given forecast horizon and output_chunk_length
+        n_pred_steps = math.ceil(n / self.output_chunk_length)
+        for cov_type, (covs, lags) in covariates.items():
+            if covs is not None:
+                relative_cov_lags[cov_type] = np.array(lags) - lags[0]
+                covariate_matrices[cov_type] = []
+                for idx, (ts, cov) in enumerate(zip(series, covs)):
+                    first_pred_ts = ts.end_time() + 1 * ts.freq
+                    last_pred_ts = (
+                        first_pred_ts
+                        + ((n_pred_steps - 1) * self.output_chunk_length) * ts.freq
+                    )
+                    first_req_ts = first_pred_ts + lags[0] * ts.freq
+                    last_req_ts = last_pred_ts + lags[-1] * ts.freq
+
+                    # check for sufficient covariate data
+                    raise_if_not(
+                        cov.start_time() <= first_req_ts
+                        and cov.end_time() >= last_req_ts,
+                        f"The corresponding {cov_type}_covariate of the series at index {idx} isn't sufficiently long. "
+                        f"Given horizon `n={n}`, `min(lags_{cov_type}_covariates)={lags[0]}`, "
+                        f"`max(lags_{cov_type}_covariates)={lags[-1]}` and "
+                        f"`output_chunk_length={self.output_chunk_length}`\n"
+                        f"the {cov_type}_covariate has to range from {first_req_ts} until {last_req_ts} (inclusive), "
+                        f"but it ranges only from {cov.start_time()} until {cov.end_time()}.",
+                    )
 
-        inference_dataset = MixedCovariatesInferenceDataset(
-            target_series=series,
-            past_covariates=past_covariates,
-            future_covariates=future_covariates,
-            n=max(
-                n + self.max_lag, n
-            ),  # required for retrieving enough future covariates
-            input_chunk_length=max(0, -self.min_lag),
-            output_chunk_length=1,
-        )
+                    if cov.has_datetime_index:
```

Comment:
Ouch, it hurts that we have to differentiate the two cases here 😱

Reply:
yes, maybe it would be nice to have a slice_inclusive in TimeSeries that always includes the start and stop for both datetime and integer indices? 🤔

Reply:
Yes maybe. I think our slicing is at least consistent with Pandas, but it's not the first time we find this a bit frustrating.